src/megatron/energon/task_encoder/base.py

```diff
@@ -229,19 +392,51 @@ def build_batch(
                 worker_config=worker_config,
             )
         elif batch_size is not None:
+            # Check if packing methods are provided by the user
+            pre_pack_provided = getattr(self.pre_pack, "__func__", None) is not TaskEncoder.pre_pack
+            final_pack_provided = (
+                getattr(self.final_pack, "__func__", None) is not TaskEncoder.final_pack
+            )
+
+            if pre_pack_provided or final_pack_provided:
+                assert (
+                    pre_pack_provided and final_pack_provided
+                ), "Both pre_pack and final_pack methods must be provided in the TaskEncoder"
+
+                assert packing_buffer_size is not None, "packing_buffer_size must be set"
+
+                dataset = PackingDataset(
+                    dataset,
+                    buffer_size=packing_buffer_size,
+                    pre_packer=self.pre_pack,
+                    final_packer=self.final_pack,
+                    final_packer_stateless=getattr(self.final_pack, "__stateless__", False),
+                    worker_config=worker_config,
+                )
+
             dataset = BatchDataset(
                 dataset,
                 batch_size=batch_size,
                 batcher=self.batch,
+                batcher_stateless=getattr(self.batch, "__stateless__", False),
                 drop_last=batch_drop_last,
                 worker_config=worker_config,
             )
+
             if getattr(self.encode_batch, "__func__", None) is not TaskEncoder.encode_batch:
-                dataset = MapDataset(dataset, self.encode_batch)
+                dataset = MapDataset(
+                    dataset,
+                    self.encode_batch,
+                    worker_config=worker_config,
+                    stateless_map_fn=getattr(self.encode_batch, "__stateless__", False),
+                )
         else:
             assert (
                 getattr(self.encode_batch, "__func__", None) is TaskEncoder.encode_batch
             ), "batch_size is not set, but encode_batch is not the default."
+            assert (
```

Comment:
`assert packing_buffer_size is None`

Reply:
Hm, actually packing should be supported even without batching I guess.